File: 10/rb/10.rb
#!/usr/bin/env ruby

arr = File.readlines('input').map(&:to_i).sort
arr.unshift 0
arr.push arr[-1]+3

# part 1
diffs = arr.each_cons(2).map{|x,y| y-x}
p diffs.count(1) * diffs.count(3)

# part 2
memo = {}
define_method 'get' do |val|
    return 0 unless arr.include? val
    return 1 if val == arr[-1]
    memo[val] ||= get(val+1) + get(val+2) + get(val+3)
end
p get 0

File: 10/jl/10.jl
#!/usr/bin/env julia

input = sort(map(x->parse(Int, x), readlines("input")))
pushfirst!(input, 0)    # well aren't you cool and different, julia
push!(input, input[end]+3)

# part 1
diffs = diff(input)
println(count(diffs .== 1) * count(diffs .== 3))

# part 2
const memo = Dict{Int,Int}()
function ways(n)
    n ∉ input       ? 0 :
    n == input[end] ? 1 :
    get!(() -> sum(ways, n+1:n+3), memo, n)
end
println(ways(0))
